File: ui/article-aem-renderer/src/main/java/org/cru/godtools/article/aem/fragment/AemArticleViewModel.java
package org.cru.godtools.article.aem.fragment;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.os.Build;
import android.view.ContextThemeWrapper;
import android.view.ViewGroup;
import android.webkit.WebView;

import com.google.common.base.Objects;

import org.cru.godtools.article.aem.R;
import org.cru.godtools.article.aem.model.Article;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

public class AemArticleViewModel extends AndroidViewModel {
    LiveData<Article> article;
    @Nullable
    private WebView mWebView;
    @NonNull
    private final ArticleWebViewClient mWebViewClient;
    @Nullable
    private String mContentUuid;

    public AemArticleViewModel(@NonNull final Application application) {
        super(application);
        mWebViewClient = new ArticleWebViewClient(application);
    }

    @NonNull
    WebView getWebView(@NonNull final Activity activity) {
        if (mWebView == null) {
            final Context context;
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
                context = new ContextThemeWrapper(activity.getApplicationContext(),
                                                  R.style.Theme_GodTools_Tool_AppBar);
            } else {
                context = new ContextThemeWrapper(activity.getApplicationContext(), activity.getTheme());
            }

            mWebView = new WebView(context);
            mWebView.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                                                                ViewGroup.LayoutParams.MATCH_PARENT));
            mWebView.setWebViewClient(mWebViewClient);
        }
        mWebViewClient.updateActivity(activity);

        return mWebView;
    }

    void updateWebViewArticle(@Nullable final Article article) {
        if (mWebView == null) {
            return;
        }
        if (article == null || article.getContent() == null) {
            return;
        }
        if (Objects.equal(article.getContentUuid(), mContentUuid)) {
            return;
        }

        mWebView.loadDataWithBaseURL(article.getUri().toString() + ".html", article.getContent(),
                "text/html", null, null);
        mContentUuid = article.getContentUuid();
    }
}

File: ui/article-aem-renderer/src/main/java/org/cru/godtools/article/aem/fragment/AemArticleViewModel.kt
package org.cru.godtools.article.aem.fragment

import android.app.Activity
import android.app.Application
import android.net.Uri
import android.os.Build
import android.view.ContextThemeWrapper
import android.view.ViewGroup
import android.webkit.WebView
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.distinctUntilChanged
import androidx.lifecycle.switchMap
import org.cru.godtools.article.aem.R
import org.cru.godtools.article.aem.db.ArticleRoomDatabase
import org.cru.godtools.article.aem.model.Article

class AemArticleViewModel(application: Application) : AndroidViewModel(application) {
    private val db = ArticleRoomDatabase.getInstance(application)

    val articleUri = MutableLiveData<Uri?>()
    private val article: LiveData<Article?> =
        articleUri.distinctUntilChanged().switchMap { db.articleDao().findLiveData(it) }

    // region WebView
    private var webView: WebView? = null
    private val webViewClient: ArticleWebViewClient = ArticleWebViewClient(application)

    fun getWebView(activity: Activity): WebView {
        webViewClient.updateActivity(activity)
        return webView ?: buildWebView(activity).also {
            webView = it
            updateWebViewArticle(article.value)
        }
    }

    private fun buildWebView(activity: Activity): WebView {
        val context = when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.M ->
                ContextThemeWrapper(activity.applicationContext, activity.theme)
            else -> ContextThemeWrapper(activity.applicationContext, R.style.Theme_GodTools_Tool_AppBar)
        }
        return WebView(context).also {
            it.layoutParams = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT
            )
            it.webViewClient = webViewClient
        }
    }
    // endregion WebView

    // region WebView Content
    private var contentUuid: String? = null

    private fun updateWebViewArticle(article: Article?) {
        if (article?.content == null) return
        if (article.contentUuid == contentUuid) return
        if (webView == null) return

        webView?.loadDataWithBaseURL("${article.uri}.html", article.content, "text/html", null, null)
        contentUuid = article.contentUuid
    }

    init {
        article.observeForever { updateWebViewArticle(it) }
    }
    // endregion WebView Content
}
